marklogic-client-api/src/main/java/com/marklogic/client/datamovement/ProgressListener.java

```diff
@@ -0,0 +1,189 @@
+
+		if (resultsSoFarWasUpdated && consumers != null) {
+			double timeSoFar = ((double) System.currentTimeMillis() - startTime) / 1000;
+
+			/**
+			 * The initial totalResults may have been incorrectly set to a value lower than jobResultsSoFar; if this
+			 * occurs, use jobResultsSoFar as the value passed to the ProgressUpdate object. totalResults is not
+			 * updated though in case there's a need to know what the initial value was.
+			 */
+			long totalForThisUpdate = jobResultsSoFar > this.totalResults && this.totalResults > 0 ? jobResultsSoFar : this.totalResults;
+
+			ProgressUpdate progressUpdate = new SimpleProgressUpdate(batch, startTime, totalForThisUpdate, timeSoFar);
+
+			for (Consumer<ProgressUpdate> consumer : consumers) {
+				invokeConsumer(consumer, progressUpdate);
+			}
+		}
+	}
+
+	/**
+	 * Protected so that a subclass can override how a consumer is invoked, particularly how an exception is handled.
+	 *
+	 * @param consumer
+	 * @param progressUpdate
+	 */
+	protected void invokeConsumer(Consumer<ProgressUpdate> consumer, ProgressUpdate progressUpdate) {
+		try {
+			consumer.accept(progressUpdate);
+		} catch (Throwable t) {
+			logger.error("Exception thrown by a Consumer<ProgressUpdate> consumer: " + consumer + "; progressUpdate: " + progressUpdate, t);
+		}
+	}
+
+	/**
+	 * Captures data of interest for a progress update.
+	 */
+	public interface ProgressUpdate {
+
+		String getProgressAsString();
+
+		boolean isComplete();
+
+		QueryBatch getQueryBatch();
+
+		long getStartTime();
+
+		long getTotalResults();
+
+		double getTimeSoFarInSeconds();
+	}
+
+	/**
+	 * Simple implementation of ProgressUpdate; only real thing of interest in here is how it generates the progress
+	 * as a string for display purposes.
+	 */
+	public static class SimpleProgressUpdate implements ProgressUpdate {
+
+		private QueryBatch queryBatch;
+		private long startTime;
+		private long totalResults;
+		private double timeSoFarInSeconds;
+
+		public SimpleProgressUpdate(QueryBatch queryBatch, long startTime, long totalResults, double timeSoFarInSeconds) {
+			this.queryBatch = queryBatch;
+			this.startTime = startTime;
+			this.timeSoFarInSeconds = timeSoFarInSeconds;
+			this.totalResults = totalResults;
+		}
+
+		@Override
+		public String getProgressAsString() {
+			if (totalResults > 0) {
+				String text = String.format("Progress: %d of %d; time %fs", queryBatch.getJobResultsSoFar(), totalResults, timeSoFarInSeconds);
+				if (timeSoFarInSeconds > 0) {
+					double rate = queryBatch.getJobResultsSoFar() / timeSoFarInSeconds;
+					BigDecimal bd = new BigDecimal(rate);
+					rate = bd.round(new MathContext(5)).doubleValue();
+					return text + "; " + rate + " records/s";
+				}
+				return text;
+			}
+
+			return String.format("Progress: %d results so far; time %fs", queryBatch.getJobResultsSoFar(), timeSoFarInSeconds);
+		}
+
+		@Override
+		public boolean isComplete() {
```

Comment:
what if the query result is 0?

Reply:
Do you mean totalResults? If that's zero, this will always return false since it doesn't know when the job completes. 

This is based on some work on a project where I first use QueryManager and a page length of zero to get a count of matching results. Then I kick off the QueryBatcher, feeding totalResults into the progress listener so it knows when "done" is arrived at.

Reply:
ah, with that design in mind this makes more sense.